### PR TITLE
Recipe app - Rails Group Capstone Project

### PR DESCRIPTION
**Pull request by gdumani**
### In this pull request based on the given layout the following functionality was implemented
- Login page and registration page:
    - Built with Devise.
- Recipes list:
    - Display a list of recipes created by the logged-in user as in the wireframe.
    - Lead to recipe details.
    - If the user is the owner of the recipe, allows the user to delete it.
- Public recipe list:
    - Displays a list of all public recipes ordered by newest as in the wireframe.
    - Leads to recipe details.
- Recipe details:
    - Displays a toggle button that allows for a recipe to be made public or private.
    - If the recipe is public or the user is the owner of the recipe, displays the recipe details as in the wireframe.
    - If the user is the owner of the recipe, leads to the form that allows the user to add new food.
    -  Counts the total food items and total price of the missing food.
- Ensured there are no N+1 queries happening.
- Navigation menu allows users to open all of the pages created.
- integration tests

### CODE REVIEW

File: app/controllers/recipe_foods_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    end
+  end
+
+  def edit
+    @recipe_food = RecipeFood.find(params[:id])
+    @recipe = Recipe.find(@recipe_food.recipe_id)
+    @food = Food.find(@recipe_food.food_id)
+    @recipe_food.update(recipe_food_params)
+    redirect_to recipes_show_path(@recipe_food.recipe_id)
+  end
```
Comment (zainsadaqat) [inline app/controllers/recipe_foods_controller.rb:23]: ### [OPTIONAL]

Overall, the code follows a standard approach commonly used in Rails applications. However, it's important to keep in mind best practices and consider potential improvements based on the specific requirements of your application. Some suggestions include:

Consider using Rails' built-in before_action or before_filter to reduce redundancy and improve code organization.
Look into using nested routes or associations to simplify the process of finding related records.
Validate the presence of required associations (recipe_id and food_id) to ensure data integrity.

File: app/controllers/shopping_list_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    @recipes = Recipe.where(user_id: current_user.id)
+    @foods = Food.where(user_id: current_user.id)
+
+    @required_foods = @recipes.flat_map do |recipe|
+      recipe.recipe_foods.includes(:food).map do |recipe_food|
+        food = recipe_food.food
+        food[:quantity] = recipe_food.quantity
+        food
+      end
+    end
```
Comment (zainsadaqat) [inline app/controllers/shopping_list_controller.rb:13]: ### [OPTIONAL]
This approach can be considered reasonable depending on the requirements of the application.

The code uses includes(:food) to eager load the associated Food records for better performance. This helps avoid the N+1 query problem by reducing the number of database queries. 👌

**Review by zainsadaqat — APPROVED** [review]
# Pull Request Approved ✅ 🎊 🎉 
## Hello @gdumani and @OsvaldoBC 👋 

### Things to appreciate
- [ ] Followed coding best practices
- [ ] Clean code
- [ ] Performed pair programming in efficient manners
- [ ] Covered all required routes

### [OPTIONAL]
- [ ] In your free time, please try to make the controllers thinner. This means only putting the relevant logic into the controller which controls the flow from controller to model or controller to view and move the logic to the models by creating methods and calling those methods in the controllers. 👍 

### Your project is complete!  There is nothing else to say other than it's time to merge it.

# Congratulations!!! 🎉 🥳 👏 🙌 🎊 ☑️ 



Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/shopping_list_controller.rb (hunk 2 above)
Comment (gdumani) [reply app/controllers/shopping_list_controller.rb:13]: Thanks for your comments and useful advice!

**Pull request merged**